pkg/controller/job/job_controller.go

```diff
@@ -1757,3 +1768,17 @@ func countReadyPods(pods []*v1.Pod) int32 {
 	}
 	return cnt
 }
+
+// This checks if we should apply PodRecreationPolicy.
+// PodRecreationPolicy controls when we recreate pods if they are marked as terminating
+// Failed means that we recreate only once the pod has terminated.
+func hasPodReplacementPolicyFailed(job *batch.Job) bool {
+	return feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) && *job.Spec.PodReplacementPolicy == batch.Failed
+}
+
+// In PodRecreationPolicy we enable counting of terminating pods.
+// We will enable this in case of JobPodReplacementPolicy being enabled or
+// PodFailurePolicy is specified.
+func shouldCountTerminating(job *batch.Job) bool {
+	return feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy) || (feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) && job.Spec.PodFailurePolicy != nil)
```

Comment:
Populating the "terminating" field solely depends on JobPodReplacementPolicy.

```suggestion
	return feature.DefaultFeatureGate.Enabled(features.JobPodReplacementPolicy)
```

Reply:
So if podfailurePolicy is specified we assume Failed but don't count the terminating field?

Reply:
I don't agree with that.  In the retriable KEP, we say that PodFailurePolicy specified is identical to the behavior as PodRecrationPolicy: `Failed` so I would assume that means that it is identical to both counting terminating pods and deletion.

Reply:
> So if podfailurePolicy is specified we assume Failed but don't count the terminating field?

That is correct. The field `terminating` is associated with the `JobPodReplacementPolicy` gate. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions

The behavior of PodFailurePolicy is something internal that doesn't necessarily need to be expressed in the API. Of course we want to expose it, but we can't do it if the feature gate that governs the field is disabled. Also note that the apiserver would wipe the field anyways (`strategy.go`)

Reply:
Sounds good.  Pushed up a change to fix that.